### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by yverenoir** [nashwaan/xml-js#21]
Hi,
I am trying to parse json to xml (both in code with `json2xml` and with cli), but it gives me this error:

```
/Users/.../node_modules/xml-js/lib/js2xml.js:35
    return (!firstLine && options.spaces ? '\n' : '') + Array(depth + 1).join(options.spaces);
                                                                         ^

RangeError: Maximum call stack size exceeded
    at Array.join (native)
    at writeIndentation (/Users/.../node_modules/xml-js/lib/js2xml.js:35:74)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:37)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
    at writeElementsCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:225:83)
    at writeElementCompact (/Users/.../node_modules/xml-js/lib/js2xml.js:204:12)
```

The original json file is:
```
{
  "vertical": {
    "-display_name": "Exercise",
    "html": { "-url_name": "12345" },
    "lti_consumer": {
      "-url_name": "12345",
      "-xblock-family": "xblock.v1",
      "-accept_grades_past_due": "false",
      "-weight": "14.0",
      "-has_score": "true",
      "-display_name": "Exercise",
      "-ask_to_send_username": "true",
      "-ask_to_send_email": "true",
      "-button_text": "Launch Exercise",
      "-custom_parameters": "none",
      "-lti_id": "id",
      "-launch_target": "new_window",
      "-launch_url": "url"
    }
  }
}
```

I would be grateful for any insights!

Thanks!
Yve

**Comment by yverenoir** [comment]
I tested the original xml file I had and converted it into json, it comes with "_attributes", after adjusting the json file, it solved the problem.

**Comment by nashwaan** [comment]
@yverenoir 
If this is solved, should this issue be closed?

**Comment by bidiu** [comment]
@yverenoir Can you please explain more about the cause of this issue here? Cause I run into the same issue. Thanks a lot ; )

**Comment by nashwaan** [comment]
@bidiu , you probably missing `"_text"` property in your json input.
For example, `{"a": "hi"}` will not produce `<a>hi</a>`. The correct input is `{"a":{"_text":"hi"}}`.

Using, the above json input provided by @yverenoir  will not work as it is missing `"_text"` property. The correct input should be:
```json
{
    "vertical": {
        "display_name": {
            "_text": "Exercise"
        },
        "html": {
            "url_name": {
                "_text": "12345"
            }
        },
        "lti_consumer": {
            "url_name": {
                "_text": "12345"
            },
            "xblock-family": {
                "_text": "xblock.v1"
            },
            "accept_grades_past_due": {
                "_text": "false"
            },
            "weight": {
                "_text": "14.0"
            },
            "has_score": {
                "_text": "true"
            },
            "display_name": {
                "_text": "Exercise"
            },
            "ask_to_send_username": {
                "_text": "true"
            },
            "ask_to_send_email": {
                "_text": "true"
            },
            "button_text": {
                "_text": "Launch Exercise"
            },
            "custom_parameters": {
                "_text": "none"
            },
            "lti_id": {
                "_text": "id"
            },
            "launch_target": {
                "_text": "new_window"
            },
            "launch_url": {
                "_text": "url"
            }
        }
    }
}
```

For more discussion on why `"_text"` property is required, see the reasons [here](https://github.com/nashwaan/xml-js/issues/20#issuecomment-309720827) and [here](https://github.com/nashwaan/xml-js/issues/19).

Please let me know if this solves the issue you are facing.

**Comment by nashwaan** [comment]
There is another reason why you might get this error:

If the text value should appear in the attributes of the xml node rather than its body, then `"_attributes"` is required. For example, `{"a": {"b": "hi"}}` will not produce `<a b="hi">`. The correct input is `{"a": "_attributes": {"b": "hi"}}}`

Thus, @yverenoir has probably used this input to solve his problem:

```json
{
    "vertical": {
        "_attributes": {
            "-display_name": "Exercise"
        },
        "html": {
            "_attributes": {
                "-url_name": "12345"
            }
        },
        "lti_consumer": {
            "_attributes": {
                "-url_name": "12345",
                "-xblock-family": "xblock.v1",
                "-accept_grades_past_due": "false",
                "-weight": "14.0",
                "-has_score": "true",
                "-display_name": "Exercise",
                "-ask_to_send_username": "true",
                "-ask_to_send_email": "true",
                "-button_text": "Launch Exercise",
                "-custom_parameters": "none",
                "-lti_id": "id",
                "-launch_target": "new_window",
                "-launch_url": "url"
            }
        }
    }
}
```

Which will produce this xml output:

```xml
<vertical -display_name="Exercise">
    <html -url_name="12345"/>
    <lti_consumer -url_name="12345" -xblock-family="xblock.v1" -accept_grades_past_due="false" -weight="14.0" -has_score="true" -display_name="Exercise" -ask_to_send_username="true" -ask_to_send_email="true" -button_text="Launch Exercise" -custom_parameters="none" -lti_id="id" -launch_target="new_window" -launch_url="url"/>
</vertical>
```

**Comment by bidiu** [comment]
@nashwaan Thank you so much for replying and following up. Yes, the cause of my problem is exactly what you said. Right now I solved it

For the record, what I am doing is to convert xml to js (compact version), and then process the js a little, and finally convert it back to xml. The problem is that I set some texts to the js directly. However, I should set the text to "_text" property instead. Thank you once again. The issue could be closed now I think.
